internal/pool/pool.go

```diff
@@ -248,15 +248,24 @@ func (p *ConnPool) getTurn() {
 	p.queue <- struct{}{}
 }
 
-func (p *ConnPool) waitTurn() error {
+func (p *ConnPool) waitTurn(ctx context.Context) error {
+	var done <-chan struct{}
+	if ctx != nil {
+		done = ctx.Done()
+	}
+
 	select {
+	case <-done:
+		return ctx.Err()
 	case p.queue <- struct{}{}:
 		return nil
 	default:
 		timer := timers.Get().(*time.Timer)
 		timer.Reset(p.opt.PoolTimeout)
 
 		select {
+		case <-done:
+			return ctx.Err()
```

Comment:
On this path, the timer isn't stopped and returned to the pool.

Reply:
Thanks